### ISSUE TITLE
certmgr --ssl https://community.letsencrypt.org hangs forever

### ISSUE THREAD

**Issue by pruiz** [mono/mono#8421]
## Steps to Reproduce

1. certmgr --ssl https://community.letsencrypt.org

Output:
```
Mono Certificate Manager - version 5.8.1.0
Manage X.509 certificates and CRL from stores.
Copyright 2002, 2003 Motus Technologies. Copyright 2004-2008 Novell. BSD licensed.
```
... and this hangs there forever.

## Current Behavior

Using mono 5.8.1 on OSX, and trying to load/access some https-enabled websites fails never returning back.

## Expected Behavior

Normal access to intended HTTPS site.

### On which platforms did you notice this

[x] macOS
[x] Linux
[ ] Windows

**Version Used**:

OSX:
```
Mono JIT compiler version 5.8.1.0 (2017-10/6bf3922f3fd Thu Mar  8 17:00:31 EST 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           normal
        SIGSEGV:       altstack
        Notification:  kqueue
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        LLVM:          yes(3.6.0svn-mono-master/8b1520c8aae)
        GC:            sgen (concurrent by default)
```

Linux (has btls enabled):
```
Mono JIT compiler version 4.8.1 (Stable 4.8.1.0/22a39d7 Sun Jul  2 06:45:50 UTC 2017)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           __thread
	SIGSEGV:       altstack
	Notifications: epoll
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug
	LLVM:          supported, not enabled.
	GC:            sgen
```

### Stacktrace

N/A

**Comment by directhex** [comment]
Confirmed on Linux, Mono 5.10.1.47

**Comment by luhenry** [comment]
```
(lldb) bt all
* thread #1, name = 'tid_307', queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff77dc6a1e libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff77f8f589 libsystem_pthread.dylib`_pthread_cond_wait + 732
    frame #2: 0x0000000102cd4b20 mono`mono_os_cond_timedwait [inlined] mono_os_cond_wait(cond=0x00007fd3a80046f8, mutex=0x00007fd3a80046b8) at mono-os-mutex.h:173 [opt]
    frame #3: 0x0000000102cd4b15 mono`mono_os_cond_timedwait(cond=0x00007fd3a80046f8, mutex=0x00007fd3a80046b8, timeout_ms=4294967295) at mono-os-mutex.c:32 [opt]
    frame #4: 0x0000000102c42ae7 mono`mono_w32handle_timedwait_signal_handle at mono-coop-mutex.h:102 [opt]
    frame #5: 0x0000000102c42ac9 mono`mono_w32handle_timedwait_signal_handle [inlined] mono_w32handle_timedwait_signal_naked(poll=0, alerted=<unavailable>) at w32handle.c:652 [opt]
    frame #6: 0x0000000102c42ac9 mono`mono_w32handle_timedwait_signal_handle(handle_data=<unavailable>, timeout=<unavailable>, poll=0, alerted=<unavailable>) at w32handle.c:767 [opt]
    frame #7: 0x0000000102c429c1 mono`mono_w32handle_wait_one(handle=<unavailable>, timeout=<unavailable>, alertable=<unavailable>) at w32handle.c:885 [opt]
    frame #8: 0x0000000102c42c32 mono`mono_w32handle_wait_multiple(handles=0x00007ffeed1fdde0, nhandles=1, waitall=0, timeout=<unavailable>, alertable=1) at w32handle.c:923 [opt]
    frame #9: 0x0000000102c20088 mono`ves_icall_System_Threading_WaitHandle_Wait_internal(handles=<unavailable>, numhandles=<unavailable>, waitall='\0', timeout=-1, error=0x00007ffeed1fdcf0) at threads.c:1982 [opt]
    frame #10: 0x00000001079cee3d
    frame #11: 0x0000000104a492e0 mscorlib.dll.dylib`System_Threading_WaitHandle_InternalWaitOne_System_Runtime_InteropServices_SafeHandle_long_bool_bool + 64
    frame #12: 0x0000000104a4928c mscorlib.dll.dylib`System_Threading_WaitHandle_WaitOne_long_bool + 44
    frame #13: 0x0000000104a490f3 mscorlib.dll.dylib`System_Threading_WaitHandle_WaitOne_int_bool + 51
    frame #14: 0x0000000104a491f1 mscorlib.dll.dylib`System_Threading_WaitHandle_WaitOne + 33
    frame #15: 0x00000001079ce83f
    frame #16: 0x00000001079c92eb
    frame #17: 0x0000000104a07e8d mscorlib.dll.dylib`System_IO_StreamWriter_Flush_bool_bool + 269
    frame #18: 0x0000000104a07d70 mscorlib.dll.dylib`System_IO_StreamWriter_Flush + 32
    frame #19: 0x0000000102fc180e
    frame #20: 0x0000000102fc0a63
    frame #21: 0x0000000102fba51b
    frame #22: 0x0000000102fba909
    frame #23: 0x0000000102a13bb7 mono`mono_jit_runtime_invoke(method=0x00000001030a8ca0, obj=<unavailable>, params=0x00007ffeed1fe518, exc=0x00000001030a9520, error=<unavailable>) at mini-runtime.c:2807 [opt]
    frame #24: 0x0000000102c01ce4 mono`do_runtime_invoke(method=0x00007fd3a7d13be8, obj=0x0000000000000000, params=0x00007ffeed1fe518, exc=0x0000000000000000, error=0x00007ffeed1fe558) at object.c:2922 [opt]
    frame #25: 0x0000000102c054d9 mono`do_exec_main_checked [inlined] mono_runtime_invoke_checked(method=<unavailable>, obj=<unavailable>, error=0x00007ffeed1fe558) at object.c:3075 [opt]
    frame #26: 0x0000000102c05498 mono`do_exec_main_checked(method=0x00007fd3a7d13be8, args=<unavailable>, error=0x00007ffeed1fe558) at object.c:4819 [opt]
    frame #27: 0x0000000102a82169 mono`mono_jit_exec(domain=<unavailable>, assembly=<unavailable>, argc=3, argv=0x00007ffeed1fe8d8) at driver.g.c:1210 [opt]
    frame #28: 0x0000000102a8514d mono`mono_main [inlined] main_thread_handler at driver.g.c:1287 [opt]
    frame #29: 0x0000000102a85119 mono`mono_main(argc=<unavailable>, argv=<unavailable>) at driver.g.c:2427 [opt]
    frame #30: 0x0000000102a032f8 mono`main [inlined] mono_main_with_options(argc=<unavailable>, argv=<unavailable>) at main.c:50 [opt]
    frame #31: 0x0000000102a032e4 mono`main(argc=4, argv=<unavailable>) at main.c:398 [opt]
    frame #32: 0x0000000102a031e4 mono`start + 52
  thread #2, name = 'SGen worker'
    frame #0: 0x00007fff77dc6a1e libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff77f8f589 libsystem_pthread.dylib`_pthread_cond_wait + 732
    frame #2: 0x0000000102cc91de mono`thread_func [inlined] mono_os_cond_wait(mutex=<unavailable>) at mono-os-mutex.h:173 [opt]
    frame #3: 0x0000000102cc91cb mono`thread_func at sgen-thread-pool.c:165 [opt]
    frame #4: 0x0000000102cc91bd mono`thread_func(data=0x0000000000000000) at sgen-thread-pool.c:196 [opt]
    frame #5: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #6: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #7: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #3, name = 'Finalizer'
    frame #0: 0x00007fff77dbd246 libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x0000000102c7701c mono`finalizer_thread [inlined] mono_os_sem_wait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:90 [opt]
    frame #2: 0x0000000102c77011 mono`finalizer_thread at mono-coop-semaphore.h:43 [opt]
    frame #3: 0x0000000102c77005 mono`finalizer_thread(unused=<unavailable>) at gc.c:893 [opt]
    frame #4: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #5: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a7d23df0) at threads.c:1123 [opt]
    frame #6: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #7: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #8: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #4
    frame #0: 0x00007fff77dc7292 libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x00007fff77f8e009 libsystem_pthread.dylib`_pthread_wqthread + 1035
    frame #2: 0x00007fff77f8dbe9 libsystem_pthread.dylib`start_wqthread + 13
  thread #5, name = 'Thread Pool I/O Selector'
    frame #0: 0x00007fff77dc6cfa libsystem_kernel.dylib`__select + 10
    frame #1: 0x0000000102cd6f27 mono`mono_poll(ufds=0x00007fd3a7c10c80, nfds=2, timeout=<unavailable>) at mono-poll.c:87 [opt]
    frame #2: 0x0000000102c291b4 mono`poll_event_wait(callback=(mono`wait_callback at threadpool-io.c:257), user_data=0x00007fd3a7f3b020) at threadpool-io-poll.c:146 [opt]
    frame #3: 0x0000000102c28a6c mono`selector_thread(data=<unavailable>) at threadpool-io.c:451 [opt]
    frame #4: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #5: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a7c10640) at threads.c:1123 [opt]
    frame #6: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #7: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #8: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #6, name = 'tid_2d03'
    frame #0: 0x00007fff77dc6a1e libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff77f8f5c2 libsystem_pthread.dylib`_pthread_cond_wait + 789
    frame #2: 0x0000000102cd4b04 mono`mono_os_cond_timedwait(cond=0x0000000102e11eb8, mutex=0x0000000102e11e78, timeout_ms=500) at mono-os-mutex.c:44 [opt]
    frame #3: 0x0000000102cdf8d9 mono`mono_thread_info_sleep at mono-coop-mutex.h:102 [opt]
    frame #4: 0x0000000102cdf8bc mono`mono_thread_info_sleep [inlined] sleep_interruptable(ms=<unavailable>, alerted=<unavailable>) at mono-threads.c:1270 [opt]
    frame #5: 0x0000000102cdf86e mono`mono_thread_info_sleep(ms=500, alerted=<unavailable>) at mono-threads.c:1302 [opt]
    frame #6: 0x0000000102b764f3 mono`monitor_thread(unused=<unavailable>) at threadpool-worker-default.c:703 [opt]
    frame #7: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #8: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a7d3e040) at threads.c:1123 [opt]
    frame #9: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #10: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #11: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #7, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a7d2a000) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #8, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a9858120) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #9, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a7f3b610) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #10, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a7c0c2c0) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #11, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a7c0c300) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #12, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a7e04e60) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #13, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a9c000f0) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #14, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a9907440) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #15, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a9e000b0) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #16, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a9d004e0) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #17, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a7d2d150) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
  thread #18, name = 'Thread Pool Worker'
    frame #0: 0x00007fff77dbd25e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x0000000102b76aa6 mono`worker_thread [inlined] mono_os_sem_timedwait(flags=MONO_SEM_FLAGS_ALERTABLE) at mono-os-semaphore.h:124 [opt]
    frame #2: 0x0000000102b76a20 mono`worker_thread at mono-coop-semaphore.h:57 [opt]
    frame #3: 0x0000000102b76a15 mono`worker_thread at threadpool-worker-default.c:391 [opt]
    frame #4: 0x0000000102b7692a mono`worker_thread(unused=<unavailable>) at threadpool-worker-default.c:483 [opt]
    frame #5: 0x0000000102c24c03 mono`start_wrapper [inlined] start_wrapper_internal at threads.c:1063 [opt]
    frame #6: 0x0000000102c24b73 mono`start_wrapper(data=0x00007fd3a985b010) at threads.c:1123 [opt]
    frame #7: 0x00007fff77f8e661 libsystem_pthread.dylib`_pthread_body + 340
    frame #8: 0x00007fff77f8e50d libsystem_pthread.dylib`_pthread_start + 377
    frame #9: 0x00007fff77f8dbf9 libsystem_pthread.dylib`thread_start + 13
```

```
Thread 1:
"<unnamed thread>" tid=0x0x7fffb02a6380 this=0x0x1052e0130 , thread handle : 0x7fd3a7d0f0f0, state : waiting
  at <unknown> <0xffffffff>
  at (wrapper managed-to-native) System.Threading.WaitHandle.Wait_internal (intptr*,int,bool,int) [0x0000b] in <f837fbf9ce4e4e0d97b332ca67afbdfb>:0
  at System.Threading.WaitHandle.WaitOneNative (System.Runtime.InteropServices.SafeHandle,uint,bool,bool) [0x00044] in <f837fbf9ce4e4e0d97b332ca67afbdfb>:0
  at System.Threading.WaitHandle.InternalWaitOne (System.Runtime.InteropServices.SafeHandle,long,bool,bool) [0x00014] in <f837fbf9ce4e4e0d97b332ca67afbdfb>:0
  at System.Threading.WaitHandle.WaitOne (long,bool) [0x00000] in <f837fbf9ce4e4e0d97b332ca67afbdfb>:0
  at System.Threading.WaitHandle.WaitOne (int,bool) [0x00019] in <f837fbf9ce4e4e0d97b332ca67afbdfb>:0
  at System.Threading.WaitHandle.WaitOne () [0x00000] in <f837fbf9ce4e4e0d97b332ca67afbdfb>:0
  at Mono.Security.Protocol.Tls.SslStreamBase.EndNegotiateHandshake (Mono.Security.Protocol.Tls.SslStreamBase/InternalAsyncResult) [0x0000e] in <4eb2484c0cc04c16a78bbcfdf31484d3>:0
  at Mono.Security.Protocol.Tls.SslStreamBase.NegotiateHandshake () [0x0002b] in <4eb2484c0cc04c16a78bbcfdf31484d3>:0
  at Mono.Security.Protocol.Tls.SslStreamBase.Write (byte[],int,int) [0x00064] in <4eb2484c0cc04c16a78bbcfdf31484d3>:0
  at System.IO.StreamWriter.Flush (bool,bool) [0x0007e] in <f837fbf9ce4e4e0d97b332ca67afbdfb>:0
  at System.IO.StreamWriter.Flush () [0x00006] in <f837fbf9ce4e4e0d97b332ca67afbdfb>:0
  at Mono.Tools.CertificateManager.GetCertificatesFromSslSession (string) [0x00087] in <0d3e2545df6547298434b6b965b640eb>:0
  at Mono.Tools.CertificateManager.Ssl (string,bool,bool) [0x0001f] in <0d3e2545df6547298434b6b965b640eb>:0
  at Mono.Tools.CertificateManager.Main (string[]) [0x001bf] in <0d3e2545df6547298434b6b965b640eb>:0
  at (wrapper runtime-invoke) <Module>.runtime_invoke_void_object (object,intptr,intptr,intptr) [0x0004e] in <0d3e2545df6547298434b6b965b640eb>:0
```

```
Mono JIT compiler version 5.12.0.90 (2018-02/91fe1eb5f58 Thu Mar  8 17:50:54 EST 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug
	Interpreter:   yes
	LLVM:          yes(3.6.0svn-mono-master/8b1520c8aae)
	GC:            sgen (concurrent by default)
```

**Comment by migueldeicaza** [comment]
This is not a simple fix.

The tool relies on both Mono's old TLS 1.0 implementation as well as poking inside the internals of the class to extract internals that are not typically surfaced.

Given that we no longer ship a managed TLS implementation, and instead have two very different implementations, we are going to need a different strategy for each backend.

**Comment by baulig** [comment]
This is actually not difficult at all.  PR incoming.

**Comment by baulig** [comment]
Should be fixed with PR #8566.
